### PR TITLE
Split Gmsh Interface into GeoLib-part and MeshLib-part

### PR DESCRIPTION
**Pull request by TomFischer**
As titled. For details see commit messages.

This is one step to resolve the issue https://github.com/ufz/ogs/issues/1164.

### CODE REVIEW

File: MeshLib/IO/GmshReader.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class Node;
+class Element;
+class Mesh;
+
+namespace IO
+{
+namespace GMSH
+{
+
+void readNodeIDs(std::ifstream& in, unsigned n_nodes,
```
Comment (norihiro-w) [inline MeshLib/IO/GmshReader.h]: Is it right that  this function should not be called from outside? If that's the case, do you still need the declaration here? same applies to the below function

File: MeshLib/IO/GmshReader.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/**
+ *
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#include <fstream>
```
Comment (norihiro-w) [inline MeshLib/IO/GmshReader.h]: maybe you can include the forward deceleration

File: MeshLib/IO/GmshReader.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/**
+ *
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#include <iosfwd>
```
Comment (norihiro-w) [inline MeshLib/IO/GmshReader.h]: no need to include this.

File: MeshLib/IO/GmshReader.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace MeshLib
+{
+class Node;
```
Comment (norihiro-w) [inline MeshLib/IO/GmshReader.h]: Forward declaration of Node and Element can be removed

File: GeoLib/IO/GMSHInterface.cpp [hunk 5]
```diff
@@ -16,12 +16,12 @@
...
 #include "BaseLib/FileTools.h"
 
-#include "FileIO/GMSHInterface.h"
-#include "FileIO/GmshIO/GMSHAdaptiveMeshDensity.h"
-#include "FileIO/GmshIO/GMSHFixedMeshDensity.h"
-#include "FileIO/GmshIO/GMSHPoint.h"
-#include "FileIO/GmshIO/GMSHPolygonTree.h"
-#include "FileIO/GmshIO/GMSHMeshDensityStrategy.h"
+#include "GeoLib/IO/GMSHInterface.h"
+#include "GeoLib/IO/GmshIO/GMSHAdaptiveMeshDensity.h"
```
Comment (chleh) [inline GeoLib/IO/GMSHInterface.cpp]: folder could be renamed to GeoLib/IO/Gmsh

File: Applications/DataExplorer/mainwindow.cpp [hunk 6]
```diff
@@ -430,10 +430,10 @@ void MainWindow::save()
         std::vector<std::string> selected_geometries;
         _project.getGEOObjects().getGeometryNames(selected_geometries);
 
-        GMSHInterface gmsh_io(
+        GeoLib::IO::GMSHInterface gmsh_io(
             _project.getGEOObjects(), true,
-            FileIO::GMSH::MeshDensityAlgorithm::AdaptiveMeshDensity,
-            0.05, 0.5, 2, selected_geometries);
+            GeoLib::IO::GMSH::MeshDensityAlgorithm::AdaptiveMeshDensity, 0.05,
```
Comment (chleh) [inline Applications/DataExplorer/mainwindow.cpp]: The directory is named GeoLib/IO/GmshIO, but the namespace is GeoLib::IO::GMSH. Maybe the two things should match?

**Comment by chleh** [comment]
:+1:

**Comment by endJunction** [comment]
I also would prefer not to repeat "IO" in the "GeoLib/IO/GmshIO" second time.

**Pull request merged**